utils/s3/client.cc

```diff
@@ -847,6 +847,82 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// a sink with known object size
+//
+// unlike upload_sink and upload_jumbo_sink, upload_ranged_sink sends the
+// received buffers right away without accumulating them first. caller is
+// supposed to put() with the payload of part_size until the specified
```

Comment:
This is too harsh requirement. Sinks are used inside output_stream-s and we don't control the way output_stream flushes its buffers

Reply:
i think we do if we want.

Reply:
Technically yes. I think that our goal is to have a sink that would pass though

```
    f = co_await open_file_dma(...)
    sz = co_await f.size()
    in = make_file_input_stream(f)
    out = make_output_stream(s3::make_upload_ranged_sink(sz))
    co_await seastar::copy(in, out) /* there's such a marvelous helper */
```

Reply:
i am not quite sure about this. will reply you once i am back from PTO.

Reply:
i think `seastar::copy()` is a recipe of poor performance. because the underlying performs deep copy for every chunk of memory read from the specified `input_stream` . see https://github.com/scylladb/seastar/blob/ce40a2fd275bd76bb752bd9a2ed99022fc3019a6/include/seastar/core/iostream-impl.hh#L538 and https://github.com/scylladb/seastar/blob/ce40a2fd275bd76bb752bd9a2ed99022fc3019a6/include/seastar/core/iostream-impl.hh#L348-L355

i think "we don't control" is the one way of control. if it's good, we can leave as it is. but if it is bad, we have to interfere with it. 

we *could* improve the performance of `seastar::copy()` though. but the point is that, i think, we can achieve an optimisitic design by fixing the different parts of system with the proper settings. in this case, with the right settings, `output_stream` can interact in the expected way with the underlying sink, and pump the memory chunk of the expected size to s3.

what is missing in this change is the implementation of `upload_ranged_sink::put(net::packet)`, which is used by `output_stream<CharType>::zero_copy_put()`.

Reply:
the constraints were dropped in the latest revision.